weave/trace_server/calls_query_builder.py

```diff
@@ -0,0 +1,726 @@
+                    path_parts.append(", " + _param_slot(pb.add_param(part), "String"))
+            safe_path = "".join(path_parts)
+            return f"(NOT (JSONType({res}{safe_path}) = 'Null' OR JSONType({res}{safe_path}) IS NULL))"
+
+    def as_select_sql(self, pb: ParamBuilder, table_alias: str) -> str:
+        if self.extra_path:
+            raise NotImplementedError(
+                "Dynamic fields cannot be selected directly, yet - implement me!"
+            )
+        return f"{super().as_sql(pb, table_alias)} AS {self.field}"
+
+    def with_path(self, path: list[str]) -> "CallsMergedDynamicField":
+        extra_path = [*(self.extra_path or [])]
+        extra_path.extend(path)
+        return CallsMergedDynamicField(
+            field=self.field, agg_fn=self.agg_fn, extra_path=extra_path
+        )
+
+    def is_heavy(self) -> bool:
+        return True
+
+
+class OrderField(BaseModel):
+    field: CallsMergedField
+    direction: typing.Literal["ASC", "DESC"]
+
+    def as_sql(self, pb: ParamBuilder, table_alias: str) -> str:
+        options: list[typing.Tuple[typing.Optional[tsi_query.CastTo], str]]
+        if isinstance(self.field, CallsMergedDynamicField):
+            # Prioritize existence, then cast to float, then str
+            options = [
+                ("exists", "desc"),
+                ("double", self.direction),
+                ("string", self.direction),
+            ]
+        else:
+            options = [(None, self.direction)]
+        res = ""
+        for index, (cast, direction) in enumerate(options):
+            if index > 0:
+                res += ", "
+            res += f"{self.field.as_sql(pb, table_alias, cast)} {direction}"
+        return res
+
+
+class Condition(BaseModel):
+    operand: "tsi_query.Operand"
+    _consumed_fields: typing.Optional[list[CallsMergedField]] = None
+
+    def as_sql(self, pb: ParamBuilder, table_alias: str) -> str:
+        conditions = process_query_to_conditions(
+            tsi_query.Query.model_validate({"$expr": {"$and": [self.operand]}}),
+            pb,
+            table_alias,
+        )
+        if self._consumed_fields is None:
+            self._consumed_fields = []
+            for field in conditions.fields_used:
+                self._consumed_fields.append(get_field_by_name(field))
+        return combine_conditions(conditions.conditions, "AND")
+
+    def _get_consumed_fields(self) -> list[CallsMergedField]:
+        if self._consumed_fields is None:
+            self.as_sql(ParamBuilder(), "calls_merged")
+        if self._consumed_fields is None:
+            raise ValueError("Consumed fields should not be None")
+        return self._consumed_fields
+
+    def is_heavy(self) -> bool:
+        for field in self._get_consumed_fields():
+            if field.is_heavy():
+                return True
+        return False
+
+
+class HardCodedFilter(BaseModel):
+    filter: tsi._CallsFilter
+
+    def is_useful(self) -> bool:
```

Comment:
Could use a comment about what it means to be useful.

Reply:
added